### DISCUSSION TITLE
[Keyball44] コンスルーを用いてビルドしたときにファームウェアの書き込みができない

### DISCUSSION THREAD

**Discussion by kk-shinoda** [Yowkees/keyball#538] [Q&A]
## 背景
Keyball44 と別途 12P 2.5mm のコンスルーを購入し、[ビルドガイド](https://github.com/Yowkees/keyball/blob/main/keyball44/doc/rev1/buildguide_jp_sub.md#build-guide-for-keyball44) を参考に組み立てを行っています。

ただ、[3.8章 ProMicroの書き込みと動作確認](https://github.com/Yowkees/keyball/blob/main/keyball44/doc/rev1/buildguide_jp_sub.md#3-8promicro%E3%81%AE%E6%9B%B8%E3%81%8D%E8%BE%BC%E3%81%BF%E3%81%A8%E5%8B%95%E4%BD%9C%E7%A2%BA%E8%AA%8D) でファームウェアの書き込みができず困っています。

### ファームウェアの書き込みができない

左右ともに同じ症状が発生しているため、はんだづけ等の問題というより何か勘違いがありそうなので相談させていただきたいです。何か誤った組み立てや確認事項に漏れがあればご教示ください。

**症状**
- PC に接続しても OLED になにも出力されない
- remap から以下のファームウェアの書き込みを試みたが、そもそも接続の候補に Keyball が上がってこない
    - Firmware: Keyball44_test_v1.3.2 for Keyball44
    - Bootloader: caterina

**確認したこと**
- PC への接続
ProMicro に USB を接続すると赤い LED が点灯するため PC への接続は問題なさそうです。

- コンスルーの向き
[ブログ記事](https://yushakobo.zendesk.com/hc/ja/articles/360044233974-%E3%82%B3%E3%83%B3%E3%82%B9%E3%83%AB%E3%83%BC-%E3%82%B9%E3%83%97%E3%83%AA%E3%83%B3%E3%82%B0%E3%83%94%E3%83%B3%E3%83%98%E3%83%83%E3%83%80-%E3%81%AE%E5%8F%96%E3%82%8A%E4%BB%98%E3%81%91%E6%96%B9%E3%82%92%E6%95%99%E3%81%88%E3%81%A6%E4%B8%8B%E3%81%95%E3%81%84) を参考にピン窓に近い方に ProMicro が来ることと、2つのコンスルーのピン窓が同じ方向(画像手前側)を向いていることを確認しました。

![IMG_5444](https://github.com/Yowkees/keyball/assets/86135975/67db5415-e0d0-44b8-a589-435d4e432281)

- シルクの場所
コンスルーと ProMicro をはんだづけした部品を基盤のシルクに合わせました。また、他方の列に装着してもうまくいかなかったためここは被疑箇所ではないと思っています。

![IMG_5447](https://github.com/Yowkees/keyball/assets/86135975/ffc24896-8d5c-41ca-96cd-eb9f4da29d75)

**Comment by kk-shinoda** [comment]
補足
書き込みの際にリセットボタンをボタンを押しましたが、keyball 44 は候補に挙がってきませんでした。

<img width="515" alt="スクリーンショット 2024-04-03 12 46 52" src="https://github.com/Yowkees/keyball/assets/86135975/defd94ec-cd16-4c19-90d3-6527b0d8a622">

参考画像 (黒塗りしている箇所は関係のないデバイスです)

**Comment by koron** [comment]
初回の書き込みは [qmk_toolbox](https://github.com/qmk/qmk_toolbox/releases/latest) か
[Pro Micro Web Updater](https://sekigon-gonnoc.github.io/promicro-web-updater/index.html) を使ってみてください。

**Reply by kk-shinoda** [reply]
@koron さん
コメントありがとうございます。
どちらも試してみましたが、うまくいきませんでした ...

PC と接続しても OLED になにも表示されないのですが、この情報は原因切り分けの役に立ちますか？

**Reply by koron** [reply]
> PC と接続しても OLED になにも表示されないのですが、この情報は原因切り分けの役に立ちますか？

残念ながらたちません。OLEDはファームウェアを書き込まないと表示されないのです。

とりあえず次はPro MicroをKeybalから外した状態で、
qmk_toolbox か Pro Micro Web Updaterを使ってファームウェアを書き込んでみてください。
Pro MicroのRSTと書かれたピンとGNDと書かれたピンを
ピンセットなどを使ってショートすることで書けるようになります。

またお使いのPro Microによっては、
書き込むためには素早く2回ショートさせる(もしくはリセットボタンを押す)
必要がある場合があります。
この時の素早くというのは、およそ0.7秒程度の猶予がありますので
さほど焦らなくても大丈夫です。

**Reply by kk-shinoda** [reply]
@koron さん
ありがとうございます。

Keyball から外して RST と GND をショートさせた状態でも書き込みの候補に挙がってきませんでした。
2回連続でショートさせるということも試みましたがだめでした。

明日遊舎工房に行くので直接相談してみようかなと思います。

**Reply by kk-shinoda** [reply]
遊舎工房に行って確認してもらったところ、ProMicro および設置方法に問題はありませんでした。

帰宅後、原因がわかり解決しましたので、その内容を記載してクローズとさせていただきます。

原因
自分の PC と Keyball の接続構成は以下のようになっています。
PC → KVMスイッチ → Keyball

ここで KVMスイッチのポートが、電力供給のみするポートだったようです。
他のポートと接続することで解決しました。

しょうもないミスでお騒がせして申し訳ありませんでした。

**Comment by kk-shinoda** [comment]
https://github.com/Yowkees/keyball/discussions/538#discussioncomment-9034432 のとおり解決しましたので、クローズします。
